### PR TITLE
[MM-38242] Directly use local http port to access Mattermost server from Apps plugin

### PR DESCRIPTION
**Pull request by hanzei**
#### Summary
This change greatly improves the performance as the apps plugin doesn't have to go through the reverse proxy any longer.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-38242

### CODE REVIEW

File: server/config/config.go [hunk 1]
```diff
@@ -107,10 +109,25 @@ func (conf *Config) Reconfigure(stored StoredConfig, mmconf *model.Config, licen
...
+	}
+	host, port, err := net.SplitHostPort(*listenAddress)
+	if err != nil {
+		return err
+	}
+
+	if host == "" {
+		host = "127.0.0.1"
+	}
+	listenURL := "http://" + host + ":" + port
```
Comment (mickmister) [inline server/config/config.go:124]: Will we run into any certificate complications when the server is configured to forward `80 -> 443`? For example, the server's SSL cert may not match the `host` we are using here. Though I could be way wrong here.

Comment (levb) [reply server/config/config.go:124]: @mickmister What certificate? (`http`, not `https`)

@hanzei nit 3/5 s/listenURL/loopbackURL/, listen(Address-not-URL) should be ":8065", on the server.

File: server/mmclient/http.go [hunk 2]
```diff
@@ -13,7 +13,7 @@ type httpClient struct {
 
 func NewHTTPClientFromSessionID(config config.Service, sessionID, actingUserID string) (Client, error) {
 	conf, mm, _ := config.Basic()
-	client, err := utils.ClientFromSession(mm, conf.MattermostSiteURL, sessionID, actingUserID)
+	client, err := utils.ClientFromSession(mm, conf.MattermostListenURL, sessionID, actingUserID)
```
Comment (levb) [inline server/mmclient/http.go:16]: ditto s/Listen/Local[host]/g

**Review by levb — CHANGES_REQUESTED** [review]
Just a naming nit, otherwise LGTM

File: server/config/config.go (hunk 1 above)
Comment (hanzei) [reply server/config/config.go]: @mickmister Good question. The `Forward80To443` setting isn't an issue as it only adds an additional listener to port 80. The caveat is in `ConnectionSecurity`. If enabled, the MM server takes care of terminating the TLS connections. In that case there using the SiteURL is preferred as indeed the host name could mismatch.

File: server/mmclient/http.go (hunk 2 above)
Comment (hanzei) [reply server/mmclient/http.go]: Good point, updated

File: server/config/config.go [hunk 3]
```diff
@@ -107,10 +109,33 @@ func (conf *Config) Reconfigure(stored StoredConfig, mmconf *model.Config, licen
...
+		host, port, err := net.SplitHostPort(*listenAddress)
+		if err != nil {
+			return err
+		}
+
+		if host == "" {
+			host = "127.0.0.1"
+		}
+
+		localURL = "http://" + host + ":" + port
```
Comment (mickmister) [inline server/config/config.go:131]: Is it possible that `port` is zero?

**Review by mickmister — APPROVED** [review]

**Review by catalintomai — APPROVED** [review]

**Review by levb — APPROVED** [review]

File: server/config/config.go (hunk 3 above)
Comment (hanzei) [reply server/config/config.go:131]: Yes, that is possible, but I would count that as a bug of the server: https://github.com/mattermost/mattermost-server/blob/92837fa1ee69c4420ea0033a35b5058ca8afc76f/model/config.go#L3692-L3702

**Review by DHaussermann — APPROVED** [review]
Tested and passed
- Tested locally by installing test app and using various commands
- Tested on an HA server by installing Service Now and ensuing the the oAutrh config functionality is working which uses an expand
No issues found

LGTM!

**Pull request merged**